### PR TITLE
Added new public api to save credentials directly

### PR DESCRIPTION
**Pull request by kreedula**
- Changed was required to support MSAL CPP adapters

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java [hunk 1]
```diff
@@ -151,6 +152,52 @@ public MsalOAuth2TokenCache(final Context context,
...
 
+    /**
+     * @param accountRecord : AccountRecord associated with the input credentials.
+     * @param credentials   : list of Credential which can include AccessTokenRecord, IdTokenRecord and RefreshTokenRecord.
+     *                      Note : Both IdTokenRecord and RefreshTokenRecord need to be non null. AccessTokenRecord can be optional.
+     * @throws ClientException : If the supplied Account or Credential are null or schema invalid.
+     */
+    public void saveCredentials(@NonNull final AccountRecord accountRecord,
+                                @NonNull final Credential... credentials) throws ClientException {
+        if (credentials == null) {
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java]: Optionally you could test if this array is `isEmpty()`

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java [hunk 2]
```diff
@@ -151,6 +152,52 @@ public MsalOAuth2TokenCache(final Context context,
         );
     }
 
+    /**
+     * @param accountRecord : AccountRecord associated with the input credentials.
+     * @param credentials   : list of Credential which can include AccessTokenRecord, IdTokenRecord and RefreshTokenRecord.
+     *                      Note : Both IdTokenRecord and RefreshTokenRecord need to be non null. AccessTokenRecord can be optional.
+     * @throws ClientException : If the supplied Account or Credential are null or schema invalid.
+     */
+    public void saveCredentials(@NonNull final AccountRecord accountRecord,
```
Comment (rpdome) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java]: I'm thinking if we should override this class and add this method only in the inherited classs (i.e. MsalCppOAuth2TokenCache).

This way, we would be able to minimize exposing this to MSAL devs.

We can mark the original saveCredentials as protected.

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java]: +1 Like this idea

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java [hunk 3]
```diff
@@ -151,6 +152,52 @@ public MsalOAuth2TokenCache(final Context context,
...
+                                @NonNull final Credential... credentials) throws ClientException {
+        if (credentials == null) {
+            throw new ClientException("Credential array passed in is null");
+        }
+
+        AccessTokenRecord accessTokenRecord = null;
```
Comment (rpdome) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java]: Should we explicitly list these three in the params?

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java (hunk 2 above)
Comment (kreedula) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java]: Thanks for the suggestion Dome. Addressed

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalCppOAuth2TokenCache.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        );
+    }
+
+    /**
+     * @param accountRecord : AccountRecord associated with the input credentials.
+     * @param credentials   : list of Credential which can include AccessTokenRecord, IdTokenRecord and RefreshTokenRecord.
+     *                      Note : Both IdTokenRecord and RefreshTokenRecord need to be non null. AccessTokenRecord can be optional.
+     * @throws ClientException : If the supplied Account or Credential are null or schema invalid.
+     */
+    public void saveCredentials(@NonNull final AccountRecord accountRecord,
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MsalCppOAuth2TokenCache.java]: @kreedula It occurs to me now that we ought to add some thread safety around this API -- if a few different threads start calling `removeRefreshTokenIfNeeded()` simultaneously we're going to have race conditions writing/reading from disk (saving new RTs while writing deleting old ones)

**Pull request merged**